mongo/datadog_checks/mongo/mongo.py

```diff
@@ -73,147 +67,62 @@ class MongoDb(AgentCheck):
-
-            username = self.instance.get('username')
-            password = self.instance.get('password')
-
-            if password and not username:
-                raise ConfigurationError('`username` must be set when a `password` is specified')
-
-            self.server = build_connection_string(
-                hosts,
-                scheme=self.instance.get('connection_scheme', 'mongodb'),
-                username=username,
-                password=password,
-                database=self.instance.get('database'),
-                options=self.instance.get('options'),
-            )
-
-        (
-            self.username,
-            self.password,
-            self.db_name,
-            self.nodelist,
-            self.clean_server_name,
-            self.auth_source,
-        ) = parse_mongo_uri(self.server, sanitize_username=bool(self.ssl_params))
-
-        self.additional_metrics = self.instance.get('additional_metrics', [])
 
         # Get the list of metrics to collect
-        self.collect_tcmalloc_metrics = 'tcmalloc' in self.additional_metrics
         self.metrics_to_collect = self._build_metric_list_to_collect()
-
-        if not self.db_name:
-            self.log.info('No MongoDB database found in URI. Defaulting to admin.')
-            self.db_name = 'admin'
-
-        # Tagging
-        custom_tags = list(set(self.instance.get('tags', [])))
-        self.service_check_tags = ["db:%s" % self.db_name] + custom_tags
-
-        # ...add the `server` tag to the metrics' tags only
-        # (it's added in the backend for service checks)
-        self.base_tags = custom_tags + ['server:%s' % self.clean_server_name]
-
-        if self.nodelist:
-            host = self.nodelist[0][0]
-            port = self.nodelist[0][1]
-            self.service_check_tags = self.service_check_tags + ["host:%s" % host, "port:%s" % port]
-
-        self.timeout = float(self.instance.get('timeout', DEFAULT_TIMEOUT)) * 1000
-
-        # Authenticate
-        self.do_auth = True
-        self.use_x509 = self.ssl_params and not self.password
-        if not self.username:
-            self.log.debug(u"A username is required to authenticate to `%s`", self.server)
-            self.do_auth = False
-
-        self.replica_check = is_affirmative(self.instance.get('replica_check', True))
-        self.collections_indexes_stats = is_affirmative(self.instance.get('collections_indexes_stats'))
-        self.coll_names = self.instance.get('collections', [])
-        self.custom_queries = self.instance.get("custom_queries", [])
-        # By default consider that this instance is a standalone, updated on each check run.
-        self.deployment = StandaloneDeployment()
         self.collectors = []
         self.last_states_by_server = {}
+        self.api = None
 
     @classmethod
     def get_library_versions(cls):
         return {'pymongo': pymongo.version}
 
     def refresh_collectors(self, mongo_version, all_dbs, tags):
```

Comment:
Change of the code logic but no behavioral change.
Instead of having all the complicated logic inside this method, this method now creates a list of potential candidates based on the config and then defer to each collector to see if it is usable or not based on the type of mongo deployment. 
This simplifies the logic in the method drastically and is less error-prone.